### PR TITLE
Auto fill file contents when new a Java file

### LINKED ISSUE #1222: Fill in package for newly created file
When creating a new file, e.g `Foo.java` inside the `com.mycomp.app`-folder then the extension should use the new `vscode.workspace.on[Will|Did]CreateFile`-api to fill in some default content, esp the `package`-declaration `package com.mycomp.app` etc.

##### Environment
- Java extension pack 0.8.1

##### Steps To Reproduce
1. Use the vscode file explorer to create a new file

##### Current Result
File is empty

##### Expected Result
File should be pre-populated

### PR DESCRIPTION
**Pull request by testforstephen**
Signed-off-by: Jinbo Wang <jinbwan@microsoft.com>

Fix #1222

- If the language server is not initialized yet, new file will only generate the class structure.
- If the language server is ready, then resolve the package name and populate both the package statement and class contents.
- If the new file is not belonged to any known source paths, then leave the package as default.

newfile

### CODE REVIEW

File: src/fileEventHandler.ts [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        const snippets: string[] = [];
+        const packageName = resolvePackageName(sourcePaths, emptyFiles[i]);
+        if (packageName) {
+            snippets.push(`package ${packageName};`);
+        }
+
+        const typeName: string = resolveTypeName(textDocuments[i].fileName);
+        snippets.push(...[
+            "",
+            `public \${1|class,interface,enum|} ${typeName} {`,
```
Comment (fbricon) [inline src/fileEventHandler.ts]: it should also propose record, if the project is compiled against Java 14 or more recent. See https://github.com/eclipse/eclipse.jdt.ls/pull/1395

File: src/fileEventHandler.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+import * as path from 'path';
+import { workspace, FileCreateEvent, ExtensionContext, FileRenameEvent, window, TextDocument, SnippetString, commands, Uri } from 'vscode';
+import { LanguageClient } from 'vscode-languageclient';
+import { ListCommandResult } from './buildpath';
+import { Commands } from './commands';
+
+let serverReady: boolean = false;
+
+export function setServerStatus(onReady: boolean) {
```
Comment (fbricon) [inline src/fileEventHandler.ts]: s/onReady/ready/g

File: src/fileEventHandler.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    serverReady = onReady;
+}
+
+export function registerFileEventHandlers(client: LanguageClient, context: ExtensionContext, ) {
+    context.subscriptions.push(workspace.onDidCreateFiles(didCreateFiles));
+}
+
+async function didCreateFiles(e: FileCreateEvent) {
```
Comment (fbricon) [inline src/fileEventHandler.ts]: s/didCreateFiles/handleNewJavaFile/g

File: src/fileEventHandler.ts (hunk 1 above)
Comment (fbricon) [reply src/fileEventHandler.ts]: I don't think this is actually doable, as, when `record` is selected, the template should be changed to insert something like `${typeName}(${0})`

Comment (testforstephen) [reply src/fileEventHandler.ts]: yes, the static snippet is not that strong. I just thought of a workaround to make the ${typeName}(${0}) to be part of the choice text, too. See the screenshot:
image

Comment (fbricon) [reply src/fileEventHandler.ts]: does it actually insert the cursor placeholder after selecting the record value?

Comment (testforstephen) [reply src/fileEventHandler.ts]: The placeholder **$0** in the choice item won't take effect.

After selected, the cursor will stop after the closing parenthesis, e.g. "public record Hello()**|** {"

**Comment by testforstephen** [comment]
PR updated.

**Comment by fbricon** [comment]
Merged as https://github.com/redhat-developer/vscode-java/commit/ce48e3fb6de6040afa9f6b81fbcbdade17237703.

Thanks @testforstephen

**Pull request closed without merging**